openedx_events/avro_attrs_bridge.py

```diff
@@ -0,0 +1,203 @@
+
+        Will also recursively generate avro records for any sub attr classes and any custom types.
+
+        Custom types are handled by AvroAttrsBridgeExtention subclass instance values defined in self.extensions dict
+        """
+        field: Dict[str, Any] = {}
+        field["name"] = field_name
+        field["type"] = dict(name=attrs_class.__name__, type="record", fields=[])
+
+        for attribute in attrs_class.__attrs_attrs__:
+            # Attribute is a simple type.
+            if attribute.type in AVRO_TYPE_FOR:
+                inner_field = {
+                    "name": attribute.name,
+                    "type": AVRO_TYPE_FOR[attribute.type],
+                }
+            # Attribute is another attrs class
+            elif hasattr(attribute.type, "__attrs_attrs__"):
+                # Inner Attrs Class
+
+                # fastavro does not allow you to redefine the same record type more than once,
+                # so only define an attr record once
+                if attribute.type.__name__ in self.schema_record_names:
+                    inner_field = {
+                        "name": attribute.name,
+                        "type": attribute.type.__name__,
+                    }
+                else:
+                    self.schema_record_names.add(attribute.type.__name__)
+                    inner_field = self.record_field_for_attrs_class(
+                        attribute.type, attribute.name
+                    )
+            else:
+                inner_field = None
+                extension = self.extensions.get(attribute.type)
+                if extension is not None:
+                    inner_field = dict(
+                        name=attribute.name, type=extension.record_fields()
+                    )
+                else:
+                    raise TypeError(
+                        f"AvroAttrsBridgeExtension for {attribute.type} not in self.exceptions."
+                    )
+            field["type"]["fields"].append(inner_field)
+
+        return field
+
+    def serialize(self, obj) -> bytes:
+        """
+        Convert from attrs to a valid avro record.
+        """
+        # Make an valid avro record from the attrs class. in the future `ID`, and time would be generated, the rest
+        # would be defined once and maybe passed into the class at instantiation time?
+        # Not sure if it makes sense to keep version info here since the schema registry will actually
+        # keep track of versions and the topic can have only one associated schema at a time.
+        obj_as_dict = attr.asdict(obj, value_serializer=self.extension_serializer)
+        # TODO what should the default values of the following be
+        avro_record = dict(
+            id="1",
+            type="Test.v1",
+            time="uea",
+            source="test_attrs",
+            sourcehost="enki",
+            minorversion=0,
+            data=obj_as_dict,
+        )
```

Comment:
For reviewer: these values are hardcoded in, I currently don't know what their eventual value should be. If you have suggestions, suggest ahead!